politeiad/cache/cockroachdb/decred.go

```diff
@@ -884,6 +884,243 @@ func (d *decred) cmdTokenInventory(payload string) (string, error) {
+func (d *decred) getVoteResultsForStartVotes(svMap map[string]StartVote) (map[string][]decredplugin.VoteOptionResult, error) {
+
+	resMap := make(map[string][]decredplugin.VoteOptionResult)
+
+	if len(svMap) == 0 {
+		return resMap, nil
+	}
+
+	tokens := make([]string, 0, len(svMap))
+	for token := range svMap {
+		tokens = append(tokens, token)
+	}
+
+	vrs := make([]VoteResults, 0, len(tokens))
+	err := d.recordsdb.
+		Where("token IN (?)", tokens).
+		Preload("Results").
+		Preload("Results.Option").
+		Find(&vrs).
+		Error
+	if err != nil {
+		return nil, err
+	}
+
+	for _, vr := range vrs {
+		resMap[vr.Token] = convertVoteOptionResultsToDecred(vr.Results)
+	}
+
+	return resMap, nil
+}
+
+// lookupResultsForVoteOptions looks in the CastVote to see how many votes
+// each option has received.
+func (d *decred) lookupResultsForVoteOptions(options []VoteOption) ([]decredplugin.VoteOptionResult, error) {
+
+	results := make([]decredplugin.VoteOptionResult, 0, len(options))
+
+	for _, v := range options {
+		var votes uint64
+		tokenVoteBit := v.Token + strconv.FormatUint(v.Bits, 16)
+		err := d.recordsdb.
+			Model(&CastVote{}).
+			Where("token_vote_bit = ?", tokenVoteBit).
+			Count(&votes).
+			Error
+		if err != nil {
+			return nil, err
+		}
+
+		results = append(results,
+			decredplugin.VoteOptionResult{
+				ID:          v.ID,
+				Description: v.Description,
+				Bits:        v.Bits,
+				Votes:       votes,
+			})
+	}
+
+	return results, nil
+}
+
+// manuallyLookupNewVoteResults manually looks up vote results for records
+// for which voting has started but there is no vote option result in resMap.
+// This is meant to be called after resMap has been populated with the records
+// for which the vote results have been lazily loaded into the VoteResults
+// table.
+func (d *decred) manuallyLookupNewVoteResults(resMap map[string][]decredplugin.VoteOptionResult, svMap map[string]StartVote) error {
+
+	for token, sv := range svMap {
+		_, ok := resMap[token]
+		if ok {
+			continue
+		}
+
+		results, err := d.lookupResultsForVoteOptions(sv.Options)
+		if err != nil {
+			return err
+		}
+
+		resMap[token] = results
+	}
+
+	return nil
+}
+
+func (d *decred) cmdBatchVoteSummary(payload string) (string, error) {
+	log.Tracef("cmdBatchVoteSummary")
+
+	bvs, err := decredplugin.DecodeBatchVoteSummary([]byte(payload))
+	if err != nil {
+		return "", err
+	}
+
+	// This query gets the latest version of each record
+	query := `SELECT a.* FROM records a
+	LEFT OUTER JOIN records b
+		ON a.token = b.token AND a.version < b.version
+	WHERE b.token IS NULL AND a.token IN (?)`
+
+	rows, err := d.recordsdb.Raw(query, bvs.Tokens).Rows()
+
+	if err != nil {
+		return "", err
+	}
+	defer rows.Close()
+
+	recordsMap := make(map[string]Record)
+	for rows.Next() {
+		var r Record
+		err := d.recordsdb.ScanRows(rows, &r)
+		if err != nil {
+			return "", err
+		}
+		recordsMap[r.Token] = r
+	}
+
+	avMap, err := d.getAuthorizeVotesForRecords(recordsMap)
```

Comment:
In some of these functions, like`getStartVotesForAuthorizeVotes` and `getVoteResultsForStartVotes`, you're passing in full maps of objects when the only thing being used is the tokens.  Wouldn't it be better to have a more generic `getStartVotes` and `getVoteResults` that take a slice of tokens instead of specific objects like a map of `AuthorizeVotes`?

Reply:
getVoteResults also needs the vote options, but otherwise yeah it makes way more sense.

Reply:
Actually, now that `getVoteResults` needs the `startVotes`, I think it makes `cmdBatchVoteSummary` look cleaner if `getStartVotes` takes the map of `AuthorizeVotes`. In this way there is a nice flow where we start with a bunch of records, a subset of which have been authorized to start voting, and a subset of which have started voting, and a subset of which have results. It's kind of weird to just extract the tokens for one of the functions. Do you agree?

Reply:
Yeah, makes sense.